packages/core/contracts/cross-chain-oracle/OracleHub.sol

```diff
@@ -0,0 +1,127 @@
+
+/**
+ * @title Gatekeeper contract deployed on mainnet that validates and sends price requests from sidechain to the DVM on
+ * mainnet. This is a "gate keeper" contract because it performs the final validation for any messages originating from
+ * a child chain's oracle before submitting price requests to the DVM. This contract also can publish DVM price
+ * resolution data to OracleSpokes on any chainId via the messenger for that chainId.
+ * @dev This contract must be a registered financial contract in order to make and query DVM price requests.
+ */
+contract OracleHub is OracleBase, Ownable, Lockable {
+    using SafeERC20 for IERC20;
+
+    // Currency that final fees are paid in.
+    IERC20 public token;
+
+    // Associates chain ID with ParentMessenger contract to use to send price resolutions to that chain's OracleSpoke
+    // contract via its ChildMessenger contract.
+    mapping(uint256 => ParentMessengerInterface) public messengers;
+
+    event SetParentMessenger(uint256 indexed chainId, address indexed parentMessenger);
+
+    constructor(address _finderAddress, IERC20 _token) OracleBase(_finderAddress) {
+        token = _token;
+    }
+
+    modifier onlyMessenger(uint256 chainId) {
+        require(msg.sender == address(messengers[chainId]), "Caller must be messenger for network");
+        _;
+    }
+
+    /**
+     * @notice Set new ParentMessenger contract for chainId.
+     * @param chainId network that has a child messenger contract that parent messenger contract will communicate with.
+     * @param messenger ParentMessenger contract that sends messages to ChildMessenger on network with ID `chainId`.
+     * @dev Only callable by the owner (presumably the Ethereum Governor contract).
+     */
+    function setMessenger(uint256 chainId, ParentMessengerInterface messenger) public nonReentrant() onlyOwner {
+        require(address(messenger) != address(0), "Invalid messenger contract");
+        messengers[chainId] = messenger;
+        emit SetParentMessenger(chainId, address(messenger));
+    }
+
+    /**
+     * @notice Publishes a DVM resolved price to the OracleSpoke deployed on the network linked  with `chainId`, or
+     * reverts if not resolved yet. This contract must be registered with the DVM to query price requests.
+     * The DVM price resolution is communicated to the OracleSpoke via the Parent-->Child messenger channel.
+     * @dev This method will return silently if already called for this price request, but will still attempt to call
+     * `messenger.sendMessageToChild` again even if it is a duplicate call. Therefore the Messenger contract for this
+     * `chainId` should determine how to handle duplicate calls.
+     * @param chainId Network to resolve price for.
+     * @param identifier Identifier of price request to resolve.
+     * @param time Timestamp of price request to resolve.
+     * @param ancillaryData extra data of price request to resolve.
+     */
+    function publishPrice(
```

Comment:
in the beginning do you think we'd like to publish a bunch of prices at once to "seed" the L2 state? probs not with any of the existing products but there could be a future situation where we'd like to do that. I think this implementation will work fine for this but it might warrant adding in `multicall` to enable this to be done more easily in a batch! we can leave this for later though, once we've thought about it a bit more.

Reply:
I don't think so at the moment. There are no existing contracts that use past prices so there is no practical need to seed the L2 state.

However I see your point about this being potentially useful in the future. I'll add in a TODO to add a multicall for this reason